python/rascaline/rascaline/utils/clebsch_gordan/_tensor_correlator.py

```diff
@@ -0,0 +1,410 @@
+        :param skip_redundant: :py:class:`bool`. If ``True``, will skip the computation
+            of redundant keys in the output tensor. This is only valid when computing an
+            auto-correlation of a density tensor. In this case, setting this to true
+            results in no loss of information, but does not preserve the norm of the
+            output tensor.
+
+        :return: A :py:class:`TensorMap` object containing the Clebsch-Gordan tensor
+            product of ``tensor_1`` and ``tensor_2``.
+        """
+
+        return self._cg_tensor_product(
+            tensor_1,
+            tensor_2,
+            o3_lambda_1_name=o3_lambda_1_name,
+            o3_lambda_2_name=o3_lambda_2_name,
+            selected_keys=selected_keys,
+            skip_redundant=skip_redundant,
+            compute_metadata=False,
+        )
+
+    def _cg_tensor_product(
+        self,
+        tensor_1: TensorMap,
+        tensor_2: TensorMap,
+        o3_lambda_1_name: str,
+        o3_lambda_2_name: str,
+        selected_keys: Optional[Labels],
+        skip_redundant: bool,
+        compute_metadata: bool,
+    ) -> TensorMap:
+        """
+        Computes the Clebsch-Gordan tensor product between ``tensor_1`` and
+        ``tensor_2``.
+
+        Executes the following steps:
+
+            1. Checks the metadata of ``tensor_1`` and ``tensor_2`` to ensure that they
+               are compatible for the Clebsch-Gordan tensor product. Ensures neither has
+               gradients (not currently supported).
+
+            2. Computes the full product of the two keys, accounting for CG combination
+               rules, and matching key dimensions that have the same name.
+
+            3. Applies the key selection to remove keys that do not need computing. In
+               the special case of a density correlation, where ``skip_redudant=True``,
+               an extra filtering of keys is applied to remove redundant keys.
+
+            4. For each key pair to be combined (and corresponding output angular
+               channel), computes the CG tensor product on the block values.
+
+            5. Returns the resulting tensor in :py:class:`TensorMap` format.
+        """
+        # 1. Check the inputs
+        self._check_inputs(tensor_1, tensor_2, selected_keys)
+
+        # 2. Compute the full product of the two keys
+        output_keys, combinations = _utils._compute_output_keys(
+            tensor_1.keys, tensor_2.keys, o3_lambda_1_name, o3_lambda_2_name
+        )
+
+        # 3. Apply key selections
+        if selected_keys is not None:
+            output_keys, combinations = _utils._apply_key_selection(
+                output_keys,
+                combinations,
+                selected_keys,
+            )
+        if skip_redundant:  # should only be used for density auto-correlations
+            output_keys, combinations = _utils._remove_redundant_keys(
+                output_keys, combinations
+            )
```

Comment:
It feels a bit weird that we are doing something in this class that is relevant for another class. I'll play with the code a bit to see if there is a way to give a "additional_key_filter" callback or similar here instead of explicitly having `skip_redundant`.

If this does not work, I can live with it!

Reply:
Thanks!